src/c/coap-server.c

```diff
@@ -0,0 +1,465 @@
+        goto finish;
+    }
+  }
+  if (!iot_data)
+  {
+    response->code = COAP_RESPONSE_CODE (400);
+    coap_add_data (response, strlen (MSG_PAYLOAD_INVALID), (uint8_t *)MSG_PAYLOAD_INVALID);
+    goto finish;
+  }
+
+  /* generate and post an event with the data */
+  devsdk_commandresult results[1];
+  results[0].origin = 0;
+  results[0].value = iot_data;
+
+  devsdk_post_readings (sdk_ctx->service, device->name, resource->name, results);
+  iot_data_free (results[0].value);
+
+  response->code = COAP_RESPONSE_CODE (204);
+
+ finish:
+  if (device)
+  {
+    edgex_free_device (device);
+  }
+}
+
+int
+run_server (coap_driver *driver)
+{
+  coap_context_t  *ctx = NULL;
+  coap_address_t bind_addr;
+  coap_resource_t *resource = NULL;
+  int result = EXIT_FAILURE;
+  sdk_ctx = driver;
+  struct sigaction sa;
+
+  coap_startup ();
+
+  /* Use EdgeX log level */
+  coap_log_t log_level;
+  switch (sdk_ctx->lc->level)
+  {
+    case IOT_LOG_ERROR:
+      log_level = LOG_ERR;
+      break;
+    case IOT_LOG_WARN:
+      log_level = LOG_WARNING;
+      break;
+    case IOT_LOG_DEBUG:
+      log_level = LOG_DEBUG;
+      break;
+    default:
+      log_level = LOG_INFO;
+  }
+  coap_set_log_level (log_level);
+  /* workaround for tinydtls log level mismatch to avoid excessive debug logging */
+  coap_tls_version_t *ver = coap_get_tls_library_version ();
+  if (ver->type == COAP_TLS_LIBRARY_TINYDTLS && log_level == LOG_INFO)
+  {
+    coap_dtls_set_log_level (log_level - 1);
+  }
+  else
+  {
+    coap_dtls_set_log_level (log_level);
+  }
+
+  /* Resolve destination address where server should be sent. Use CoAP default ports. */
+  coap_proto_t proto = COAP_PROTO_UDP;
+  char *port = "5683";
+  if (driver->security_mode != SECURITY_MODE_NOSEC)
+  {
+    proto = COAP_PROTO_DTLS;
+    port = "5684";
+  }
+  if (resolve_address (iot_data_string (driver->coap_bind_addr), port, &bind_addr) < 0) {
+    iot_log_error (sdk_ctx->lc, "failed to resolve CoAP bind address");
+    goto finish;
+  }
+
+  /* setup libcoap for a server */
+  if (!(ctx = coap_new_context (NULL)))
+  {
+    iot_log_error (sdk_ctx->lc, "cannot initialize context");
+    goto finish;
+  }
+
+  if (driver->security_mode == SECURITY_MODE_PSK)
+  {
+    /* use iterator just to get address of PSK key data */
+    iot_data_array_iter_t array_iter;
+    iot_data_array_iter (driver->psk_key, &array_iter);
+    iot_data_array_iter_next(&array_iter);
+
+    if (!(coap_context_set_psk (ctx, "", (uint8_t *)iot_data_array_iter_value (&array_iter),
+                                iot_data_array_length (driver->psk_key))))
+    {
+      iot_log_error (sdk_ctx->lc, "cannot initialize PSK");
+      goto finish;
+    }
+  }
+
+  if (!(coap_new_endpoint (ctx, &bind_addr, proto)))
+  {
+    iot_log_error (sdk_ctx->lc, "cannot initialize listen endpoint");
+    goto finish;
+  }
+
+  /* Creates handler for PUT, which is not what we want... */
+  resource = coap_resource_unknown_init (&data_handler);
```

Comment:
coap_resource_init is not needed? Please confirm

Reply:
That is correct, not needed in device-coap. `coap_resource_init()` means you know the precise path for the resource. In our case, the path is in the form `/a1r/{deviceName}/{resourceName}`. We look up deviceName dynamically in `parse_path()` with `edgex_get_device_byname()`, and then validate the resource from the list in the device's profile.

libcoap _does_ support the RESTful approach of using a PUT to an unknown resource to trigger generation of a libcoap resource via`coap_resource_unknown_init()` and `coap_resource_init()`. See libcoap's [coap-server example](https://github.com/obgm/libcoap/blob/f681525272c1ab2d1abbeed0cfb03edba23d8936/examples/coap-server.c#L1392) for this kind of use. However, for device-coap this approach would mean two sources of truth for device/resource -- EdgeX and libcoap -- which must be coordinated. The current approach is simple and follows how device-rest-go works.

If devices and resources do not change often, then this coordination might be worthwhile. IMO this approach is future work though.